src/Internal/Service/DuplicatePaymentPreventionService.php

```diff
@@ -0,0 +1,220 @@
+<?php
+/**
+ * Class DuplicatePaymentPreventionService
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Service;
+
+use Exception;
+use WC_Order;
+use WCPay\Constants\Intent_Status;
+use WCPay\Core\Server\Request\Get_Intention;
+
+/**
+ * Used for methods, which detect existing payments or payment intents,
+ * and prevent creating duplicate payments.
+ */
+class DuplicatePaymentPreventionService {
+	/**
+	 * Key name for saving the current processing order_id to WC Session with the purpose
+	 * of preventing duplicate payments in a single order.
+	 *
+	 * @type string
+	 */
+	const SESSION_KEY_PROCESSING_ORDER = 'wcpay_processing_order';
+
+	/**
+	 * Flag to indicate that a previous order with the same cart content has already paid.
+	 *
+	 * @type string
+	 */
+	const FLAG_PREVIOUS_ORDER_PAID = 'wcpay_paid_for_previous_order';
+
+	/**
+	 * Flag to indicate that a previous intention attached to the order was successful.
+	 */
+	const FLAG_PREVIOUS_SUCCESSFUL_INTENT = 'wcpay_previous_successful_intent';
+
+	/**
+	 * OrderService instance.
+	 *
+	 * @var OrderService
+	 */
+	private $order_service;
+
+	/**
+	 * Initializes all dependencies and hooks, related to the service.
+	 *
+	 * @param OrderService $order_service The order service instance.
+	 */
+	public function init( OrderService $order_service ) {
```

Comment:
1. The order service should be provided in the class constructor.
2. `init` should be `init_hooks`, and we can add it in `class WC_Payments` for now.

Reply:
😀 Good point to notice. Anyway, I am still in the progress of working on this one. I will take care of this one.

Reply:
Updated the code but there is a new concern
https://github.com/Automattic/woocommerce-payments/pull/7450/files#r1356675850